aggregation_server/server.py

```diff
@@ -0,0 +1,266 @@
+        """
+        with self.global_lock:
+            if request.request_id in self.aggregation_contexts:
+                logger.warning(
+                    f"[CONFIGURE] Request context already exists for request_id='{request.request_id}'"
+                )
+                return ConfigureResponse(
+                    status="Already configured for this request_id"
+                )
+            self.aggregation_contexts[request.request_id] = AggregationContext(
+                request.request_id, request.num_of_workers
+            )
+
+        logger.info(
+            f"[CONFIGURE] Created AggregationContext for request_id='{request.request_id}' "
+            f"with expected workers: {request.num_of_workers}"
+        )
+        return ConfigureResponse(status="Configured")
+
+    def Aggregate(self, request, context):
+        """
+        Process an aggregation response from a worker:
+          1. Retrieve the AggregationContext.
+          2. Append the worker’s response (and trigger the computation if this is the last response).
+          3. Wait until the result is ready.
+          4. Finalize the aggregation and return the result.
+        """
+        agg_ctx = self._get_aggregation_context(request.request_id, context)
+        self._append_response_and_compute(agg_ctx, request, context)
+        self._wait_for_result(agg_ctx, request, context)
+        result = self._finalize_and_get_result(agg_ctx, context)
+        return AggregateResponse(result=result)
+
+    def Cleanup(self, request, context):
+        """
+        Remove the AggregationContext for this request_id.
+        (Useful if the request is complete and no further aggregations are expected.)
+        """
+        with self.global_lock:
+            if request.request_id in self.aggregation_contexts:
+                del self.aggregation_contexts[request.request_id]
+                logger.info(
+                    f"[CLEANUP] Removed AggregationContext for request_id='{request.request_id}'"
+                )
+                return CleanupResponse(status="Cleaned up")
+            else:
+                logger.warning(
+                    f"[CLEANUP] No AggregationContext found for request_id='{request.request_id}'"
+                )
+                return CleanupResponse(status="No aggregation found for request_id")
+
+    def _get_aggregation_context(self, request_id, context):
+        """
+        Retrieve the AggregationContext associated with the given request_id.
+        Aborts the RPC if no such context exists.
+        """
+        with self.global_lock:
+            if request_id not in self.aggregation_contexts:
+                msg = f"Request ID '{request_id}' not configured."
+                logger.error(f"[AGGREGATE] {msg}")
+                context.abort(grpc.StatusCode.INVALID_ARGUMENT, msg)
+            return self.aggregation_contexts[request_id]
+
+    def _append_response_and_compute(self, agg_ctx, request, context):
```

Comment:
Split this in 2 separate methods.

`_append_response` and `_compute_result_if_ready` or something similar